File: Functional-Stack.cpp
#include <iostream>
#include <memory>
#include <vector>
#include <chrono>

using namespace std;
using namespace chrono;

// Node structure to mimic immutability
struct Node {
    int value;
    shared_ptr<Node> next;

    Node(int val, shared_ptr<Node> nxt = nullptr) : value(val), next(nxt) {}
};

// Functional Stack class
class FunctionalStack {
private:
    shared_ptr<Node> head;

public:
    FunctionalStack() : head(nullptr) {}

    // Push an element onto the stack
    FunctionalStack push(int val) const {
        return FunctionalStack(make_shared<Node>(val, head));
    }

    // Pop an element from the stack
    pair<int, FunctionalStack> pop() const {
        if (!head) {
            throw runtime_error("Pop from empty stack");
        }
        return {head->value, FunctionalStack(head->next)};
    }

    // Peek at the top element
    int peek() const {
        if (!head) {
            throw runtime_error("Peek from empty stack");
        }
        return head->value;
    }

    // Check if the stack is empty
    bool isEmpty() const {
        return head == nullptr;
    }

    // Constructor for creating a stack with a specific head
    FunctionalStack(shared_ptr<Node> newHead) : head(newHead) {}
};

// Utility function to measure execution time
template <typename Func>
void timeOp(const string& label, Func&& func) {
    auto start = high_resolution_clock::now();
    func();
    auto end = high_resolution_clock::now();
    auto duration = duration_cast<nanoseconds>(end - start).count();
    cout << label << " execution time: " << duration << " ns" << endl;
}

// Test function for the stack
void testFunctionalStack() {
    FunctionalStack stack;

    // Measure initial empty stack
    timeOp("Check empty stack", [&]() { stack.isEmpty(); });

    // Push elements
    FunctionalStack currentStack = stack;
    timeOp("Push 100000 elements", [&]() {
        for (int i = 0; i < 100000; ++i) {
            currentStack = currentStack.push(i);
        }
    });

    // Peek the top element
    timeOp("Peek at top", [&]() { cout << "Top of stack after pushing 100000 elements: " << currentStack.peek() << endl; });

    // Pop elements
    timeOp("Pop 100000 elements", [&]() {
        for (int i = 0; i < 100000; ++i) {
            currentStack = currentStack.pop().second;
        }
    });

    // Check empty stack
    timeOp("Final check for empty stack", [&]() { currentStack.isEmpty(); });
}

int main() {
    testFunctionalStack();
    return 0;
}

File: Functional-Stack.hs
import System.CPUTime
import Text.Printf

-- Define the Functional Stack as a list
data Stack a = Stack [a]
 deriving (Show)

-- Create an empty stack
emptyStack :: Stack a
emptyStack = Stack []

-- Push an element onto the stack
push :: a -> Stack a -> Stack a
push x (Stack xs) = Stack (x : xs)

-- Pop an element from the stack
pop :: Stack a -> (Maybe a, Stack a)
pop (Stack [])     = (Nothing, Stack []) -- Handle empty stack
pop (Stack (x:xs)) = (Just x, Stack xs)

-- Peek the top element without modifying the stack
peek :: Stack a -> Maybe a
peek (Stack [])     = Nothing
peek (Stack (x:_))  = Just x

-- Check if the stack is empty
isEmpty :: Stack a -> Bool
isEmpty (Stack []) = True
isEmpty _          = False

-- Time a single operation with detailed output
timeOp :: String -> IO a -> IO a
timeOp label action = do
    start <- getCPUTime
    result <- action
    end <- getCPUTime
    let diffNano :: Integer
        diffNano = end - start -- Difference in picoseconds
    let diffMicro :: Double
        diffMicro = fromIntegral diffNano / (10^6) -- Convert to microseconds
    let diffMilli :: Double
        diffMilli = fromIntegral diffNano / (10^9) -- Convert to milliseconds
    let diffSeconds :: Double
        diffSeconds = fromIntegral diffNano / (10^12) -- Convert to seconds
    printf "%s execution time: %.0f ns (%.3f µs, %.6f ms, %.9f s)\n"
        label (fromIntegral diffNano :: Double) diffMicro diffMilli diffSeconds
    return result

-- Testing the Functional Stack
testStack :: IO ()
testStack = do
    -- Create an empty stack
    let stack = emptyStack :: Stack Int -- Specify type as Stack Int
    putStrLn $ "Initial stack: " ++ show stack

    -- Time the push operations
    stack1 <- timeOp "Push 10" $ return (push 10 stack)
    stack2 <- timeOp "Push 20" $ return (push 20 stack1)
    stack3 <- timeOp "Push 30" $ return (push 30 stack2)
    putStrLn $ "After pushing 30: " ++ show stack3

    -- Time the pop operations
    (top1, stack4) <- timeOp "Pop" $ return (pop stack3)
    putStrLn $ "Popped item: " ++ show top1 ++ ", Remaining stack: " ++ show stack4

    (top2, stack5) <- timeOp "Pop again" $ return (pop stack4)
    putStrLn $ "Popped item: " ++ show top2 ++ ", Remaining stack: " ++ show stack5

    -- Time the peek operation
    top3 <- timeOp "Peek" $ return (peek stack5)
    putStrLn $ "Top item: " ++ show top3 ++ ", Stack remains: " ++ show stack5

    -- Check if the stack is empty
    putStrLn $ "Is stack empty? " ++ show (isEmpty stack)  -- True
    putStrLn $ "Is stack5 empty? " ++ show (isEmpty stack5) -- False

    -- Attempt to pop from an empty stack
    (top4, stack6) <- timeOp "Pop from empty stack" $ return (pop (emptyStack :: Stack Int))
    putStrLn $ "Popped item from empty stack: " ++ show top4 ++ ", Remaining stack: " ++ show stack6
